### PR TITLE
[ADD] advance_check_void

### PR DESCRIPTION
**Pull request by patrickrwilson**
New module that adds history for voiding checks.

### CODE REVIEW

File: advance_check_void/README.rst [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+To use this module
+
+* Applicable to payments that are type Check only
+* Go to Accounting > Vendors > Payments
+* Go to the payment that has payment type Check and Check is printed
+* There are two options of Void
+* Void: Voids check only and allows to reprint check
+* Void Check and Payment: Reverses payment on date specified, Opens bills, Reconciles AP and Bank accounts and log void details
+* The payment is now in Void status for Audit purposed
```
Comment (dreispt) [inline advance_check_void/README.rst]:
```suggestion
* The payment is now in Void status for Audit purposes
```

File: advance_check_void/__manifest__.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+{
+    "name": "Advance Check Void",
+    "version": "14.0.1.0.0",
+    "category": "Accounting",
+    "summary": "Void History for Check Payments",
+    "author": "Open Source Integrators",
+    "website": "https://github.com/ursais/osi-addons",
+    "license": "AGPL-3",
+    "depends": [
+        "account_accountant",
```
Comment (dreispt) [inline advance_check_void/__manifest__.py]: Can we lower the dependency requirements?
```suggestion
        "account",
```

File: advance_check_void/__manifest__.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        "l10n_us_check_printing",
```
Comment (dreispt) [inline advance_check_void/__manifest__.py]: Is US check printing really a requirement?
```suggestion
```

File: advance_check_void/models/account_payment.py [hunk 4]
```diff
@@ -0,0 +1,391 @@
...
+                    rec.is_visible_check = True
+                    sequence = rec.journal_id.check_sequence_id
+                    sequence.next_by_id()
+                    message = (
+                        """<ul class="o_mail_thread_message_tracking">
+                        <li>(_(Check Updated Date:)) %s</li>
+                        <li>(_(Check Number:)) %s (Generated)</li>
+                        <li>(_(State:)) %s</li></ul>"""
+                    ) % (
+                        datetime.now().strftime(DEFAULT_SERVER_DATE_FORMAT),
```
Comment (dreispt) [inline advance_check_void/models/account_payment.py]: This produces server timezone date, you probably want [user timezone date](https://www.odoo.com/documentation/14.0/developer/reference/addons/orm.html#odoo.fields.Datetime.context_timestamp).
```suggestion
                        fields.Date.to_string(fields.Date.context_timezone(self)),
```

File: advance_check_void/models/account_payment.py [hunk 5]
```diff
@@ -0,0 +1,391 @@
+# Copyright (C) 2019 - TODAY, Open Source Integrators
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from datetime import datetime
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+from odoo.tools import DEFAULT_SERVER_DATE_FORMAT
```
Comment (dreispt) [inline advance_check_void/models/account_payment.py]: You should not need this.
Instead is [fields.Date methods](https://www.odoo.com/documentation/14.0/developer/reference/addons/orm.html#odoo.fields.Date).
```suggestion
```

File: advance_check_void/models/account_payment.py [hunk 6]
```diff
@@ -0,0 +1,391 @@
...
+            elif (
+                self.payment_method_id.payment_type == "outbound"
+                and self.payment_method_code in ("check_printing", "ACH-Out", "manual")
+            ):
+                self.is_visible_check = True
+                if not self.check_manual_sequencing:
+                    self.is_readonly_check = True
+
+    def print_checks(self):
+        payment_check_void_obj = self.env["payment.check.void"]
```
Comment (dreispt) [inline advance_check_void/models/account_payment.py]: Suggestion to use Object naming conventions, instead of old API 
```suggestion
        PaymentCheckVoid = self.env["payment.check.void"]
```

File: advance_check_void/models/account_payment.py [hunk 7]
```diff
@@ -0,0 +1,391 @@
...
+        }
+
+    def write(self, vals):
+        if vals.get("check_number") and not str(vals.get("check_number")).isdigit():
+            raise ValidationError(_("Check number must be integer."))
+        result = super(AccountPayment, self).write(vals)
+        check_hist_obj = self.env["payment.check.history"]
+        for res in self:
+            new_chk_id = False
+            self._cr.execute(
```
Comment (dreispt) [inline advance_check_void/models/account_payment.py]: Can we replace this SQL with an ORM `write()`?
```suggestion
            self._cr.execute(
```

File: advance_check_void/models/account_payment.py [hunk 8]
```diff
@@ -0,0 +1,391 @@
...
+                    "name": res.name,
+                    "payment_id": res.id,
+                    "partner_id": res.partner_id.id,
+                    "amount": res.amount,
+                    "check_amount": res.amount,
+                    "check_number": res.check_number,
+                    "journal_id": res.journal_id.id,
+                    "date": fields.Date.context_today(res),
+                    "create_date": fields.Date.context_today(res),
+                    "write_date": fields.Date.context_today(res),
```
Comment (dreispt) [inline advance_check_void/models/account_payment.py]: This should be timespamps, not simple dates
```suggestion
                    "create_date": fields.Date.now(),
                    "write_date": fields.Date.now(),
```

File: advance_check_void/models/account_payment.py [hunk 9]
```diff
@@ -0,0 +1,391 @@
...
+                    "create_uid": res.create_uid.id,
+                    "state": res.state,
+                    "is_visible_check": True if not res.check_number else False,
```
Comment (dreispt) [inline advance_check_void/models/account_payment.py]: Let's write this in a Pyhtonic way
```suggestion
                    "is_visible_check": not res.check_number,
```

File: advance_check_void/models/account_payment.py [hunk 10]
```diff
@@ -0,0 +1,391 @@
...
+                check_ids = check_hist_obj.search(
+                    [("payment_id", "=", res.id), ("write_date", "<=", res.write_date)],
+                    order="id desc",
+                    limit=1,
+                )
+            if check_ids:
+                for chk in check_ids:
+                    if res.state == "sent":
+                        continue
+                    chk.write({"state": "void"})
```
Comment (dreispt) [inline advance_check_void/models/account_payment.py]: Why the additonal line?
```suggestion
                    if res.state != "sent":
                        chk.write({"state": "void"})
```

File: advance_check_void/models/account_payment.py [hunk 11]
```diff
@@ -0,0 +1,391 @@
...
+                and record.check_number <= 0
+            ):
+                raise ValidationError(_("Please Enter Check Number"))
+        return res
+
+
+class PaymentCheckHistory(models.Model):
+    _name = "payment.check.history"
+    _description = "Payment Info for the check Payment Feature"
+    _order = "write_date desc"
```
Comment (dreispt) [inline advance_check_void/models/account_payment.py]: You set `write_date desc` default order, but then force `id desc` on every search. Why is that?
Should it be `id desc` in the first place?

File: advance_check_void/wizard/account_move_reversal.py [hunk 12]
```diff
@@ -0,0 +1,113 @@
+# Copyright (C) 2019 - TODAY, Open Source Integrators
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+import logging
+from datetime import datetime
+
+from odoo import _, api, fields, models
+from odoo.exceptions import ValidationError
+from odoo.tools import DEFAULT_SERVER_DATE_FORMAT
```
Comment (dreispt) [inline advance_check_void/wizard/account_move_reversal.py]: The field.Date functions so that this can be removed

File: advance_check_void/wizard/account_move_reversal.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+
+_logger = logging.getLogger(__name__)
+
+
+class AccountMoveReversal(models.TransientModel):
+    """Account move reversal wizard, it cancel an account move by reversing it."""
+
```
Comment (dreispt) [inline advance_check_void/wizard/account_move_reversal.py]: When inheriting, don't reset the Class description
```suggestion
```

File: advance_check_void/wizard/print_prenumbers_checks.py [hunk 14]
```diff
@@ -0,0 +1,46 @@
+# Copyright (C) 2019 - TODAY, Open Source Integrators
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from datetime import datetime
+
+from odoo import _, models
+from odoo.exceptions import ValidationError
+from odoo.tools import DEFAULT_SERVER_DATE_FORMAT
```
Comment (dreispt) [inline advance_check_void/wizard/print_prenumbers_checks.py]: To be removed.

File: advance_check_void/wizard/print_prenumbers_checks.py [hunk 15]
```diff
@@ -0,0 +1,46 @@
...
+
+
+class PrintPreNumberedChecks(models.TransientModel):
+    _inherit = "print.prenumbered.checks"
```
Comment (dreispt) [inline advance_check_void/wizard/print_prenumbers_checks.py]: This does not seem related to voiding checks, as advertised in the module title.
Maybe it is a different feature and a different module?

**Review by dreispt — COMMENTED** [review]
Added suggestions for improvement

File: advance_check_void/wizard/print_prenumbers_checks.py (hunk 15 above)
Comment (patrickrwilson) [reply advance_check_void/wizard/print_prenumbers_checks.py]: I agree, this can be removed from this module.

Comment (dreispt) [reply advance_check_void/wizard/print_prenumbers_checks.py]: Maybe make it a second module, to not lose the feature.

Comment (patrickrwilson) [reply advance_check_void/wizard/print_prenumbers_checks.py]: We can do that, this module was migrated from a previous customer but others are wanting to use it so we are making it more generalized for other projects. I'm guessing this piece was a specific feature for that customer as well as the check report changes which were already removed.

**Comment by patrickrwilson** [comment]
@Freni-OSI excellent work, can you make pre-commit happy and squash commits please.

**Comment by patrickrwilson** [comment]
@dreispt your suggested changes should be applied, can you review again?

File: advance_check_void/README.rst [hunk 16]
```diff
@@ -0,0 +1,76 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+    :alt: License: AGPL-3
+
+========
+Overview
+========
+
+This module is used to make vendor bill payments via checks and also void the checks. 
+With this module we can keep a track of payment check History and also see all voided Checks.
```
Comment (dreispt) [inline advance_check_void/README.rst:10]: The README should be automatically generated.
Please add the fragments in a README subdirectory.

File: advance_check_void/models/account_payment.py [hunk 17]
```diff
@@ -0,0 +1,381 @@
...
+                    "check_number": res.check_number,
+                    "journal_id": res.journal_id.id,
+                    "date": fields.Datetime.now(),
+                    "create_date": fields.Datetime.now(),
+                    "write_date": fields.Date.context_today(res),
+                    "create_uid": res.create_uid.id,
+                    "state": res.state,
+                    "is_visible_check": not res.check_number,
+                }
+                new_chk_id = check_hist_obj.create(check_hist)
```
Comment (dreispt) [inline advance_check_void/models/account_payment.py]: Tip: variables with recordset values should not end with `_id`

File: advance_check_void/README.rst (hunk 16 above)
Comment (dreispt) [reply advance_check_void/README.rst:10]: ...maybe not in the OSI repo.

**Review by dreispt — APPROVED** [review]

File: advance_check_void/README.rst (hunk 16 above)
Comment (patrickrwilson) [reply advance_check_void/README.rst:10]: I'll add it anyway just to have it in there, that way if the repo is ever updated to generate them or it goes to OCA it will already be done.

**Review by omahadik — APPROVED** [review]
